--- qca/plugins/qca-ossl/CMakeLists.txt.underlink~	2014-07-16 10:56:57.873308236 +0200
+++ qca/plugins/qca-ossl/CMakeLists.txt	2014-07-16 10:57:24.224978795 +0200
@@ -35,9 +35,7 @@ if(OPENSSL_FOUND)
   target_link_libraries(qca-ossl ${QCA_LIB_NAME})
   target_link_libraries(qca-ossl ${OPENSSL_LIBRARIES})
 
-  if(APPLE)
-    target_link_libraries(qca-ossl crypto)
-  endif(APPLE)
+  target_link_libraries(qca-ossl crypto)
 
   if(WIN32)
     add_definitions(-DOSSL_097)
