--- qca/CMakeLists.txt.bero	2012-12-27 16:00:31.466535869 +0100
+++ qca/CMakeLists.txt	2012-12-27 16:00:41.515414799 +0100
@@ -142,7 +142,7 @@ endif(NOT WIN32)
 configure_file("crypto.prf.cmake" "${CMAKE_BINARY_DIR}/crypto.prf" @ONLY)
 install(FILES "${CMAKE_BINARY_DIR}/crypto.prf" DESTINATION ${CMAKE_INSTALL_PREFIX}/mkspecs/features)
 
-install(FILES man/qcatool2.1 DESTINATION ${CMAKE_INSTALL_PREFIX}/share/man/man1)
+install(FILES man/qcatool2.1 DESTINATION /usr/share/man/man1)
 
 add_subdirectory(src)
 add_subdirectory(tools)
